lib/llm/src/block_manager/metrics_kvbm.rs

```diff
@@ -57,24 +66,179 @@ impl KvbmMetrics {
+
+            if let Err(err) = axum::serve(listener, app)
+                .with_graceful_shutdown(async move {
+                    // wait for shutdown signal
+                    notify_for_task.notified().await;
+                })
+                .await
+            {
+                tracing::error!("[kvbm] metrics server error: {err}");
+            }
+        };
+
+        // Spawn on existing runtime if present, otherwise start our own.
+        if tokio::runtime::Handle::try_current().is_ok() {
+            tokio::spawn(run_server);
+        } else {
+            thread::spawn(move || {
+                let rt = tokio::runtime::Builder::new_multi_thread()
+                    .enable_all()
+                    .build()
+                    .expect("build tokio runtime");
+                rt.block_on(run_server);
+            });
+        }
+
         Self {
             offload_requests,
             offload_blocks_d2h,
             onboard_requests,
             onboard_blocks_h2d,
             onboard_blocks_d2d,
-            save_kv_layer_requests,
             matched_tokens,
+            shutdown_notify: Some(notify),
+        }
+    }
+}
+
+impl Drop for KvbmMetrics {
+    fn drop(&mut self) {
+        if let Some(n) = &self.shutdown_notify {
+            // (all KvbmMetrics clones) + 1 (held by server task)
+            // strong_count == 2 means this is the last metrics instance
+            if Arc::strong_count(n) == 2 {
+                n.notify_waiters();
+            }
+        }
+    }
+}
+
+/// GET /metrics
+async fn metrics_handler(State(registry): State<Arc<Registry>>) -> Response {
+    let metric_families = registry.gather();
+    let encoder = TextEncoder::new();
+
+    let mut buf = Vec::new();
+    if let Err(e) = encoder.encode(&metric_families, &mut buf) {
+        let mut resp = Response::new(Body::from(format!("encode error: {e}")));
+        *resp.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+        resp.headers_mut().insert(
+            header::CONTENT_TYPE,
+            HeaderValue::from_static("text/plain; charset=utf-8"),
+        );
+        return resp;
+    }
+
+    let mut resp = Response::new(Body::from(buf));
+    *resp.status_mut() = StatusCode::OK;
+    resp.headers_mut().insert(
+        header::CONTENT_TYPE,
+        HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
+    );
+    resp
+}
+
+/// A raw, standalone Prometheus metrics registry implementation using the fixed prefix: `kvbm_`
+#[derive(Debug, Clone)]
+pub struct KvbmMetricsRegistry {
+    registry: Arc<Registry>,
+    prefix: String,
+}
+
+impl KvbmMetricsRegistry {
+    pub fn new() -> Self {
+        Self {
+            registry: Arc::new(Registry::new()),
+            prefix: "kvbm".to_string(),
         }
     }
+
+    pub fn create_intcounter(
+        &self,
+        name: &str,
+        description: &str,
+        labels: &[(&str, &str)],
+    ) -> anyhow::Result<IntCounter> {
```

Comment:
For my own understanding, why does this function return an `anyhow::Result` instead of just a normal `Result`?